sympy/sets/tests/test_setexpr.py

```diff
@@ -0,0 +1,180 @@
+from sympy.sets.setexpr import SetExpr
+from sympy.utilities.pytest import XFAIL
+from sympy.sets import Interval, FiniteSet
+from sympy import Expr, Set, exp, log, sin, cos, Symbol, Min, Max, S, oo
+
+I = Interval(0, 2)
+
+
+def test_setexpr():
+    se = SetExpr(Interval(0, 1))
+    assert isinstance(se.set, Set)
+    assert isinstance(se, Expr)
+
+
+def test_scalar_funcs():
+    assert SetExpr(Interval(0, 1)).set == Interval(0, 1)
+    a, b = Symbol('a', real=True), Symbol('b', real=True)
+    a, b = 1, 2
+    # TODO: add support for more functions in the future:
+    for f in [exp, log]:
+        input_se = f(SetExpr(Interval(a, b)))
+        output = input_se.set
+        expected = Interval(Min(f(a), f(b)), Max(f(a), f(b)))
+        assert output == expected
+
+
+def test_Add_Mul():
+    assert (SetExpr(Interval(0, 1)) + 1).set == Interval(1, 2)
+    assert (SetExpr(Interval(0, 1)) * 2).set == Interval(0, 2)
+
+
+def test_Pow():
+    assert (SetExpr(Interval(0, 2))**2).set == Interval(0, 4)
+
+
+def test_compound():
+    assert (exp(SetExpr(Interval(0, 1)) * 2 + 1)).set == \
+        Interval(exp(1), exp(3))
+
+
+def test_Interval_Interval():
+    assert (SetExpr(Interval(1, 2)) + SetExpr(Interval(10, 20))).set == \
+        Interval(11, 22)
+    assert (SetExpr(Interval(1, 2)) * SetExpr(Interval(10, 20))).set == \
+        Interval(10, 40)
+
+
+def test_FiniteSet_FiniteSet():
+    assert (SetExpr(FiniteSet(1, 2, 3)) + SetExpr(FiniteSet(1, 2))).set ==\
+        FiniteSet(2, 3, 4, 5)
+    assert (SetExpr(FiniteSet(1, 2, 3)) * SetExpr(FiniteSet(1, 2))).set ==\
+        FiniteSet(1, 2, 3, 4, 6)
+
+
+def test_Interval_FiniteSet():
+    assert (SetExpr(FiniteSet(1, 2)) + SetExpr(Interval(0, 10))).set == \
+        Interval(1, 12)
+
+
+def test_Many_Sets():
+    assert (SetExpr(Interval(0, 1)) +
+                    SetExpr(Interval(2, 3)) +
+                    SetExpr(FiniteSet(10, 11, 12))).set == Interval(12, 16)
+
+
+def test_same_setexprs_are_not_identical():
+    a = SetExpr(FiniteSet(0, 1))
+    b = SetExpr(FiniteSet(0, 1))
+    assert (a + b).set == FiniteSet(0, 1, 2)
+
+    # Cannont detect the set being the same:
+    # assert (a + a).set == FiniteSet(0, 2)
```

Comment:
> So which of the two does it do? The second one is the most correct.

It's not implemented like that. This expression is commented out. The correctness is discutibile, one may reason for both ways.

Reply:
In order to have a notion of sets being the "same" or not, you have to have to include something else alongside the set. Equal sets are equal. That's both mathematically true and how SymPy works. 

The difference here is between `{x - y | x in a, y in a}` and the diagonal `{x - x | x in a, x in a}`. The diagonal is obviously contained in the general case (subset). For full generality, the set expression construction should use a different dummy variable for each operand. 

It won't be possible to get a wrong result from doing it this way, although you may get something that is larger than you wanted. The original motivation here is to do interval arithmetic. If f(x) = sin(x) and g(x) = cos(x) both are contained in the interval [-1, 1], then it is not wrong that f - g and f - f are contained in  [-2, 2] (actually neither case is tight). However, it is wrong if we give {0} in the f - g case. The issue of course is that if we let a = [-1, 1], there is nothing tying a to f or to g, so without that information we have to assume that each instance of a came from a different function.

The primary issue here is that SymPy doesn't have a way to distinguish two identical sets. Perhaps a clearer solution would be to add a name to a set expression. This would make it similar to the random variable case, where X - X is 0 but X - Y is not, even if X and Y are iid. Perhaps this would clear up the printing issue I mentioned as well.

We could also go at the interval arithmetic directly and associate an arbitrary expression with a SetExpr, with `SetExpr(expr, set)` meaning that `expr` is bounded by `set`. Two SetExprs are not equal unless they have the same set and expr. This is just an idea (not sure yet if it's a good one).

I'm really curious how the original code attempted to fix this issue.

Reply:
> I'm really curious how the original code attempted to fix this issue.

@mrocklin had a counter variable registered in the _args_ of `SetExpr`. At this point, with the same set and the same counter, `a - a == FiniteSet(0)` for any `a = SetExpr(...)`. That is, if both _set_ and _counter_ match, build `{x - x | x in a}`, otherwise build `{x - y | x in a, y in a}`.

The alternative of using a name for `SetExpr` as it is done in `RandomSymbol` could also work, but maybe it's too cumbersome.

Otherwise we could ignore the _diagonal_ set construction, as always assume that the elements are drawn from two sets (which is the case in this PR).